### PR TITLE
LP-118: Review Apps

### PR DESCRIPTION
**Pull request by Polynya**
## Include a summary of what this merge request involves (*)
Create a [review app](https://docs.gitlab.com/ee/ci/review_apps/) for each pull request, i.e. an environment for testing a feature or bugfix before it is merged. The URL has the pattern https://portal---BRANCH-NAME.livelyforest-609fad32.uksouth.azurecontainerapps.io 

BRANCH_NAME is a sanitised version of the branch name e.g. _feature-lp-118_ for branch _feature/LP-118_

This PR contains 3 Github workflows. The 2 for events triggered by opening and closing PRs are working successfully, as seen by this pull request.
The third workflow is triggered by comments (/start, /stop, /cleanup) but will only run on the develop branch so cannot be tested yet. When the processes have been tested on new features, we may decide to remove pr_open so that the developer can choose when the code is ready for a review app.

Starting a review app includes these steps:
- Clone the Dev database (_drupal_portal_) to _drupal_ci_BRANCH-NAME_
- Deactivate any revisions that have been left running
- Create new revision of Container App
- Add label BRANCH_NAME to the revision
- Run `drush deploy`

A review app is removed when the PR is closed. The steps include:
- remove the label from the revision
- deactivate any revisions without a label or any traffic. 
- delete the database

It's not clear yet whether the Dev environment will always be correct (running off the develop branch) but it isn't really needed now. Only the database is needed so that it can be cloned for each reviewapp.

## Call out any relevant implementation decisions
- Are there any links to back up your chosen methodology?
  - https://dev.to/zirkelc/trigger-github-workflow-for-comment-on-pull-request-45l2

- Why have you taken the approach?
- Any particular problem areas?
SSO is not implemented yet. It depends on infrastructure changes which are in progress.

There are sleeps in a couple of places, which I will try to replace later.

Later I would like to remove some of the duplication between workflows, perhaps using composite actions: https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

## This PR has been tested in the following browsers
- [ ] Arc
- [ ] Edge
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] IE 11 (Windows)
- [ ] iOS Chrome
- [ ] iOS Safari
- [ ] Android Chrome
- [ ] Android Firefox
- [ ] Android default

### CODE REVIEW

**Comment by Polynya** [comment]
reviewapp

**Comment by Polynya** [comment]
start

**Comment by Polynya** [comment]
start

**Comment by Polynya** [comment]
Hopefully /stop will work once this PR has been merged. It has to be in the develop branch so I haven't been able to test it yet

**Comment by siliconmeadow** [comment]
> Hopefully /stop will work once this PR has been merged. It has to be in the develop branch so I haven't been able to test it yet

Aha! You knew exactly what I was checking :-)

File: .github/workflows/open_pr.yml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          REF_NAME: ${{ github.head_ref }}
+        shell: bash
+        run: |
+          CI_COMMIT_BRANCH="$REF_NAME"
+          ENV_SLUG=$(echo "${REF_NAME}" | sed -e 's:[^[:alpha:]|^[:digit:]]:-:g' | sed -e 's/\(.*\)/\L\1/')
+          tag=":$ENV_SLUG"
+          az extension add --name containerapp --allow-preview true
+          az containerapp job start -n portal-dbclone -g rg-ecc-portal-uks-dev --subscription "Essex County Council (Portal)" --image "acreccuksdev.azurecr.io/dbclone" --env-vars "NEW_DB_SUFFIX=${ENV_SLUG}" 'MYSQL_HOST=mariadb-ecc-uks-dev.mariadb.database.azure.com' 'MYSQL_USER=mariadb-root' 'MYSQL_DATABASE=drupal_portal' 'MYSQL_PASSWORD=secretref:mysql-password' --container-name dbclonea --cpu 0.75 --memory 1.5
+          echo "Running on branch '$CI_COMMIT_BRANCH': tag = $tag"
+          sleep 4m
```
Comment (Polynya) [inline .github/workflows/open_pr.yml:37]: There's a 4 minute wait which I've left in from Matt's work. I'll try to find a better way later.

File: .github/workflows/close_pr.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+      - name: Remove label
+        id: remove_label
+        env:
+          ARM_CLIENT_SECRET: ${{ secrets.ARM_CLIENT_SECRET }}
+          REF_NAME: ${{ github.head_ref }}
+        run: |
+          ENV_SLUG=$(echo "${REF_NAME}" | sed -e 's:[^[:alpha:]|^[:digit:]]:-:g' | sed -e 's/\(.*\)/\L\1/')
+          az containerapp revision label remove --label ${ENV_SLUG} --name portal -g rg-ecc-portal-uks-dev --subscription "Essex County Council (Portal)"
+          sleep 1m
```
Comment (Polynya) [inline .github/workflows/close_pr.yml:32]: This sleep is here because I found that the next job started too quickly and didn't deactivate the revision.

**Review by siliconmeadow — APPROVED** [review]

**Review by MatthewWilkes — APPROVED** [review]

**Pull request merged**